File: Objective-C/Reference App/210. Waypoints List UI/DisplayWaypointsListViewController.m
//
//  DisplayWaypointsListViewController.m
//  Reference App
//
//  Created by David Deller on 8/3/18.
//  Copyright © 2018 TallyGo. All rights reserved.
//

#import "DisplayWaypointsListViewController.h"

@interface DisplayWaypointsListViewController ()

@end

@implementation DisplayWaypointsListViewController

- (IBAction)go:(id)sender {
    [self getRoute];
}

- (void)getRoute {
    // Get these coordinates from your app, these are just a sample
    NSArray<TGWaypoint *> *waypoints = @[
        [TGWaypoint.alloc initWithCoordinate:CLLocationCoordinate2DMake(34.101558, -118.340944) address:nil description:@"Grauman's Chinese Theatre"],
        [TGWaypoint.alloc initWithCoordinate:CLLocationCoordinate2DMake(34.07902875, -118.379441) address:nil description:@"Quarter point"],
        [TGWaypoint.alloc initWithCoordinate:CLLocationCoordinate2DMake(34.0564995, -118.417938) address:nil description:@"Midpoint"],
        [TGWaypoint.alloc initWithCoordinate:CLLocationCoordinate2DMake(34.03397025, -118.456435) address:nil description:@"Three-quarters point"],
        [TGWaypoint.alloc initWithCoordinate:CLLocationCoordinate2DMake(34.011441, -118.494932) address:nil description:@"Santa Monica Pier"],
    ];
    
    TGRouteRequest *request = [TGRouteRequest.alloc initWithWaypoints:waypoints];
    
    [TGNavigationService routeForRequest:request completionHandler:^(TGRouteResponse * _Nonnull response) {
        if (response.error != nil) {
            // handle error
            
            [self handleError:response.error];
            
        } else {
            TGRoute *route = response.route;
            // do something with route
            
            [self showWaypointsList:route];
        }
    }];
    
}

- (void)showWaypointsList:(TGRoute *)route {
    TGWaypointsListViewController *viewController = [TGWaypointsListViewController create];
    viewController.route = route;
    
    viewController.nextWaypoint = route.segments.firstObject.destinationWaypoint;
    viewController.getCurrentLocationWaypoint = ^TGWaypoint * _Nonnull{
        return route.segments.firstObject.originWaypoint;
    };
    
    if (self.navigationController != nil) {
        self.navigationController.navigationBar.hidden = NO;
        [self.navigationController pushViewController:viewController animated:YES];
        
    } else {
        viewController.navigationItem.leftBarButtonItem = [UIBarButtonItem.alloc initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(pressDoneButton:)];
        [self presentViewController:[viewController embedInNavigationController] animated:YES completion:nil];
    }
}

// This method isn't necessary if you're pushing it onto a UINavigationController
- (void)pressDoneButton:(UIBarButtonItem *)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

@end

File: Swift 4/Reference App/210. Waypoints List UI/DisplayWaypointsListViewController.swift
//
//  DisplayWaypointsListViewController.swift
//  Reference App
//
//  Created by David Deller on 6/19/18.
//  Copyright © 2018 TallyGo. All rights reserved.
//

import UIKit
import TallyGoKit

class DisplayWaypointsListViewController: ExampleViewController {

    @IBAction func go(_ sender: Any) {
        getRoute()
    }
    
    // MARK: -
    
    func getRoute() {
        // Get these coordinates from your app, these are just a sample
        let waypoints: [TGWaypoint] = [
            TGWaypoint(coordinate: CLLocationCoordinate2D(latitude: 34.101558, longitude: -118.340944), address: nil, description: "Grauman's Chinese Theatre"),
            TGWaypoint(coordinate: CLLocationCoordinate2D(latitude: 34.07902875, longitude: -118.379441), address: nil, description: "Quarter point"),
            TGWaypoint(coordinate: CLLocationCoordinate2D(latitude: 34.0564995, longitude: -118.417938), address: nil, description: "Midpoint"),
            TGWaypoint(coordinate: CLLocationCoordinate2D(latitude: 34.03397025, longitude: -118.456435), address: nil, description: "Three-quarters point"),
            TGWaypoint(coordinate: CLLocationCoordinate2D(latitude: 34.011441, longitude: -118.494932), address: nil, description: "Santa Monica Pier"),
        ]
        
        let request = TGRouteRequest(waypoints: waypoints)
        
        TGNavigationService.route(for: request) { (response) in
            if let error = response.error {
                // handle error
                
                self.handleError(error)
                
            } else if let route = response.route {
                // do something with route
                
                self.showWaypointsList(route: route)
            }
        }
    }
    
    func showWaypointsList(route: TGRoute) {
        let viewController = TGWaypointsListViewController.create()
        viewController.route = route
        
        viewController.nextWaypoint = route.segments.first!.destinationWaypoint
        viewController.getCurrentLocationWaypoint = { () -> TGWaypoint in
            return route.segments.first!.originWaypoint
        }
        
        if let navigationController = navigationController {
            navigationController.navigationBar.isHidden = false
            navigationController.pushViewController(viewController, animated: true)
            
        } else {
            viewController.navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(pressDoneButton(_:)))
            present(viewController.embedInNavigationController(), animated: true, completion: nil)
        }
    }
    
    // This method isn't necessary if you're pushing it onto a UINavigationController
    @objc func pressDoneButton(_ sender: UIBarButtonItem) {
        dismiss(animated: true, completion: nil)
    }

}
